shared/src/main/java/org/jetbrains/skija/Surface.java

```diff
@@ -4,37 +4,388 @@
+     *
+     * @see <a href="https://fiddle.skia.org/c/@Surface_notifyContentWillChange">https://fiddle.skia.org/c/@Surface_notifyContentWillChange</a>
+     */
+    public void notifyContentWillChange(ContentChangeMode mode) {
+        Stats.onNativeCall();
+        _nNotifyContentWillChange(_ptr, mode.ordinal());
+    }
+
+    /**
+     * <p>Returns the recording context being used by the Surface.</p>
+     *
+     * @return the recording context, if available; null otherwise
+     */
+    @Nullable
+    public DirectContext getRecordingContext() {
+        Stats.onNativeCall();
+        long ptr = _nGetRecordingContext(_ptr);
+        return ptr == 0 ? null : new DirectContext(ptr);
+    }
+
+    /**
+     * <p>Returns Canvas that draws into Surface.</p>
+     *
+     * <p>Subsequent calls return the same Canvas.
+     * Canvas returned is managed and owned by Surface, and is deleted when Surface is deleted.</p>
+     *
+     * @return Canvas for Surface
+     */
+    @NotNull
     public Canvas getCanvas() {
         Stats.onNativeCall();
-        return new Canvas(_nGetCanvas(_ptr), false, this);
+        long ptr = _nGetCanvas(_ptr);
+        return ptr == 0 ? null : new Canvas(ptr, false, this);
     }
 
+    /**
+     * <p>Returns a compatible Surface, or null.</p>
+     *
+     * <p>Returned Surface contains the same raster, GPU, or null properties as the original.
+     * Returned Surface does not share the same pixels.</p>
+     *
+     * <p>Returns null if imageInfo width or height are zero, or if imageInfo is incompatible with Surface.</p>
+     *
+     * @param imageInfo contains width, height, AlphaType, ColorType, ColorSpace
+     * @return compatible SkSurface or null
+     * @see <a href="https://fiddle.skia.org/c/@Surface_makeSurface">https://fiddle.skia.org/c/@Surface_makeSurface</a>
+     */
+    @Nullable
+    public Surface makeSurface(ImageInfo imageInfo) {
+        Stats.onNativeCall();
+        long ptr = _nMakeSurfaceI(_ptr,
+                imageInfo._width,
+                imageInfo._height,
+                imageInfo._colorInfo._colorType.ordinal(),
+                imageInfo._colorInfo._alphaType.ordinal(),
+                Native.getPtr(imageInfo._colorInfo._colorSpace));
+        return new Surface(ptr);
+    }
+
+    /**
+     * <p>Calls makeSurface(ImageInfo) with the same ImageInfo as this surface,
+     * but with the specified width and height.</p>
+     *
+     * <p>Returned Surface contains the same raster, GPU, or null properties as the original.
+     * Returned Surface does not share the same pixels.</p>
+     *
+     * <p>Returns null if imageInfo width or height are zero, or if imageInfo is incompatible with Surface.</p>
+     *
+     * @param width  pixel column count; must be greater than zero
+     * @param height pixel row count; must be greater than zero
+     * @return compatible SkSurface or null
+     */
+    @Nullable
+    public Surface makeSurface(int width, int height) {
```

Comment:
@tonsky Do you think we should keep this or it's better to remove it and keep only `Surface makeSurface(ImageInfo imageInfo)`? The doc for `Surface makeSurface(int width, int height)` says that it calls the other one under the hoods but with different width and height. What do you think?

Reply:
I think yes. In general, if something is in Skia just for convenience, it might be a good idea to have it in Skija as well.

As an optional optimization, if multiple C++ methods are implemented in terms of one another, sometimes it’s easier to impement corresponding java methods with single native method, just varying the parameters.